### PR TITLE
Fixed problem with accessing backend in TestDslLazyLoader#method_missing.

### LINKED ISSUE #4524: Dynamic resource loading fails in certain RSpec extension blocks
# Describe the problem

The dynamic resource DSL-loading code introduced in #4365 altered the way that the resource DSL is imported into the control code space. User who use RSpec features like `before`/`after` blocks and `shared_example` blocks, which are not documented or supported Chef InSpec features, may experience errors if they reference dynamically loaded resources within such blocks.

First seen publicly in release 4.17.7 of Chef InSpec.

## Reproduction 

```
control "repro-1" do
  describe 'resource-in-before' do
    before(:all) do
      command('true')
    end
    it "required to do something" do
    end
  end
end
```

output:
```
Profile: tests from foo.rb (tests from foo.rb)
Version: (not specified)
Target:  local://

  ×  repro-1: resource-in-before
     ×  resource-in-before required to do something
     undefined method `command' for #<RSpec::ExampleGroups::ResourceInBefore before(:context) hook>

```

### LINKED ISSUE #4587: inspec 4.18.0 breaks `subject`-style tests
# Describe the problem
After upgrading this code breaks:

```
describe service('redis-server') do
  it 'is listening on port 6379' do
    expect(port(6379)).to be_listening
  end
end
```

```
     ×  is listening on port 6379
     undefined method `port' for #<RSpec::ExampleGroups::ServiceRedisServer_2:0x00007fc75d89b350>
```

This works with 4.17.17

## Possible Solution
?

### PR DESCRIPTION
**Pull request by zenspider**
Fixes #4587.
Fixes #4524.

Signed-off-by: Ryan Davis <zenspider@chef.io>

### CODE REVIEW

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Please add tests that exercise both bugs. In these cases, I find it clearest to simply add a functional test with a simple profile that clearly demonstrates the reproduction case.

We can later iterate on improving those tests (perhaps to move them out of the "functionals"), but for now confidence is key.

**Comment by miah** [comment]
I'm +1 for tests, -1 for functionals. They are too slow.

**Comment by zenspider** [comment]
They are too slow. I just wrote the equivalent functional and unit and the functional test is literally 5x slower. That's on a mac w/ SSD. It would be horrifically slower on windows.

File: test/unit/runner_test.rb [hunk 1]
```diff
@@ -3,9 +3,20 @@
...
 require "inspec/secrets"
 require "inspec/runner"
+require "fetchers/mock"
 
 describe Inspec::Runner do
-  let(:runner) { Inspec::Runner.new({ command_runner: :generic }) }
+  let(:runner) { Inspec::Runner.new({ command_runner: :generic, reporter: [] }) }
+
+  it "bug #4524" do
+    file = 'describe "x" do before(:all) { command("y") }; it "z" do end; end'
```
Comment (clintoncwolfe) [inline test/unit/runner_test.rb]: Please, for clarity, expand this into a heredoc, or place in an external file and make it look like more typical control code. I can help you with that if you would like.

A major advantage of having the control code in an external file is that it is easy to run it using `inspec exec` during testing and development.

I want to build trust and confidence with our users by showing that code that looks exactly like theirs works in our tests.

File: test/unit/runner_test.rb [hunk 2]
```diff
@@ -3,9 +3,20 @@
...
+    runner.add_target("bug4524.rb" => file)
+    runner.load
+
+    result = RSpec::Core::Runner.new(nil).run_specs(runner.test_collector.tests)
```
Comment (clintoncwolfe) [inline test/unit/runner_test.rb:21]: I suggest that these three lines be combined into a helper method.

File: test/unit/runner_test.rb (hunk 1 above)
Comment (clintoncwolfe) [reply test/unit/runner_test.rb]: I am not sure this test belongs in this file - perhaps a new file, runner_dsl_test.rb or similar. 

Once established, and helpers created, there are many functional tests that could likely be migrated to such a unit test file, which would be faster.

File: test/unit/runner_test.rb [hunk 3]
```diff
@@ -3,9 +3,20 @@
...
 require "inspec/runner"
+require "fetchers/mock"
 
 describe Inspec::Runner do
-  let(:runner) { Inspec::Runner.new({ command_runner: :generic }) }
+  let(:runner) { Inspec::Runner.new({ command_runner: :generic, reporter: [] }) }
+
+  it "bug #4524" do
+    file = 'describe "x" do before(:all) { command("y") }; it "z" do end; end'
+    runner.add_target("bug4524.rb" => file)
```
Comment (miah) [inline test/unit/runner_test.rb:18]: Think you forgot to add this file?

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
I am glad that you have found a way to make this work using the much faster unit test approach, and your concern for the developers' test speed is addressed.

I'm still looking for test coverage - the test given covers the `before:` issue but does not provide a test case for the `subject` issue that was reported. 

I don't have full confidence in the unit testing mockup system here. I suggest a compromise in which the profile code under test is stored externally in files, able to be tested by the unit testing system but also able to be accessed manually for diagnosis and confirmation.

**Comment by clintoncwolfe** [comment]
As this is a regression, and for the sake of getting this into a release, all I'm really concerned about for this is adding a test case for `subject`. If a tech debt issue were created to move the test fixtures into files and re-arrange / generalize the test setup, I'd be OK with this moving ahead.

File: test/unit/runner_test.rb (hunk 3 above)
Comment (zenspider) [reply test/unit/runner_test.rb:18]: It's the "mock" way of fetching. Totally undocumented but that's a separate issue.

**Review by miah — APPROVED** [review]

**Pull request merged**